### PR TITLE
5375 allow adding directories as Logical File Set

### PR DESCRIPTION
**Pull request by jkho**

### CODE REVIEW

**Comment by APriestman** [comment]
Please merge in develop and fix the merge conflicts.

File: Core/src/org/sleuthkit/autopsy/casemodule/AddLocalFilesTask.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
  * case database, grouped under a virtual directory that serves as the data
  * source.
  */
-class AddLocalFilesTask implements Runnable {
+public class AddLocalFilesTask implements Runnable {
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/casemodule/AddLocalFilesTask.java]: Why are you changing anything in this class?

File: Core/src/org/sleuthkit/autopsy/logicalimager/configuration/LogicalImagerConfig.java [hunk 2]
```diff
@@ -71,18 +77,21 @@
...
         this.promptBeforeExit = true;
+        this.createVHD = false;
         this.ruleSets = ruleSets;
     }
 
     LogicalImagerConfig(
         String version,
         boolean finalizeImageWriter,
         boolean promptBeforeExit,
+        boolean creatVHD,
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/logicalimager/configuration/LogicalImagerConfig.java]: Typo

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java [hunk 3]
```diff
@@ -182,70 +274,171 @@ private String addReport(Path reportPath, String reportName) {
...
+                            Long.parseLong(crtime),
+                            Long.parseLong(atime),
+                            Long.parseLong(mtime),
+                            localFilesDataSource);
+
+                    lineNumber++;
+                } // end reading file
+
+                trans.commit();
+                newDataSources.add(localFilesDataSource);
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: I think logically these should be after the next paren, to keep the commit at the same level as the start of the transaction.

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java [hunk 4]
```diff
@@ -182,70 +274,171 @@ private String addReport(Path reportPath, String reportName) {
...
+            }
+        } catch (TskCoreException ex) {
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: This needs to catch all the possible exceptions so we ensure the transaction gets closed.

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java [hunk 5]
```diff
@@ -182,70 +274,171 @@ private String addReport(Path reportPath, String reportName) {
...
-            try {
-                // index the artifact for keyword search
-                blackboard.indexArtifact(artifact);
-            } catch (Blackboard.BlackboardException ex) {
-                LOGGER.log(Level.SEVERE, "Unable to index blackboard artifact " + artifact.getArtifactID(), ex); //NON-NLS
+            artifacts.add(artifact);
+        }
+    }
+
+    private void addExtractedFiles(File src, Path resultsPath, List<Content> newDataSources) throws TskCoreException, UnsupportedEncodingException, IOException {
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: This doesn't appear to be cancellable?

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java [hunk 6]
```diff
@@ -182,70 +274,171 @@ private String addReport(Path reportPath, String reportName) {
...
         return imagePathToObjIdMap;
     }
-    
+
     @Messages({
         "# {0} - line number", "# {1} - fields length", "# {2} - expected length", "AddLogicalImageTask.notEnoughFields=File does not contain enough fields at line {0}, got {1}, expecting {2}",
         "# {0} - target image path", "AddLogicalImageTask.cannotFindDataSourceObjId=Cannot find obj_id in tsk_image_names for {0}"
     })
-    private void addInterestingFiles(File src, Path resultsPath) throws IOException, TskCoreException {
+    private void addInterestingFiles(File src, Path resultsPath, boolean createVHD) throws IOException, TskCoreException {
```
Comment (APriestman) [inline Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: This doesn't appear to be cancellable?

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java (hunk 5 above)
Comment (jkho) [reply Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: Do you want to rollback the transaction upon cancel?

Comment (APriestman) [reply Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: Yes

File: Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java (hunk 6 above)
Comment (jkho) [reply Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: I was able to detect the cancel and return, but I don't think there is a rollback for interesting items.

Comment (APriestman) [reply Core/src/org/sleuthkit/autopsy/logicalimager/dsp/AddLogicalImageTask.java]: No I made a note in the testing story that I don't think we can rollback at this point.

File: Core/src/org/sleuthkit/autopsy/datamodel/utils/LocalFileImporter.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+     * @param crtime      Created time
+     * @param atime       Access time
+     * @param mtime       Modified time
+     * @param dataSource  The data source to add the file to
+     * 
+     * @return The AbstractFile that was just created
+     * 
+     * @throws TskCoreException 
+     */
+    public AbstractFile addLocalFile(File fileOnDisk, String nameInImage, String pathInImage,
```
[the remark this answers is not in the document]
Comment (jkho) [reply Core/src/org/sleuthkit/autopsy/datamodel/utils/LocalFileImporter.java]: I changed to name and parentPath

File: Core/src/org/sleuthkit/autopsy/datamodel/utils/LocalFileImporter.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+     * it and its parents as needed.
+     * 
+     * @param file       The file to get the SpecialDirectory for
+     * @param dataSource The data source
+     * 
+     * @return The SpecialDirectory object corresponding to the given file
+     * 
+     * @throws TskCoreException 
+     */
+    private SpecialDirectory getLocalFilesDir(File file, Content dataSource) throws TskCoreException {
```
[the remark this answers is not in the document]
Comment (jkho) [reply Core/src/org/sleuthkit/autopsy/datamodel/utils/LocalFileImporter.java]: Renamed method to getOrMakeDirInDataSource
Changed parameter File file to File directory

**Pull request merged**